### PR TITLE
(New) Add option to generate additional debug info for expression dereferencing pointer to pointers

### PR DESCRIPTION
**Pull request by huangjd**
This is a different implementation to #94100, which has been reverted.

When -fdebug-info-for-profiling is specified, for any Load expression if the pointer operand is not a declared variable, clang will emit debug info describing the type of the pointer operand (which can be an intermediate expr)

### CODE REVIEW

**Comment by huangjd** [comment]
It seems the restriction for dbg.value appearing in the front end was lifted recently, so now I can just emit dbg.value directly

**Comment by pogo59** [comment]
You should have a look at [this imminent patch](https://github.com/llvm/llvm-project/pull/91724) which will affect your test.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (jmorse) [inline clang/lib/CodeGen/CGDebugInfo.cpp:5787]: Part of the move towards debug-info away from intrinsics involves not inserting with raw instruction pointers, but instead iterators [0]. In practice I believe all that needs to be done is using `InsertBB->begin()` to fetch an iterator, you'll likely need to take `end()` and check it below for the empty-block case.

I understand #94224 provides a more graceful solution by allowing insertion at end(), however that won't land for a short while.

[0] https://discourse.llvm.org/t/rfc-instruction-api-changes-needed-to-eliminate-debug-intrinsics-from-ir/68939

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (jmorse) [inline clang/lib/CodeGen/CGDebugInfo.cpp:5765]:
```suggestion
  // case we have nothing to do. Clang emits declared variable as alloca, and
```
```suggestion
  // case we have nothing to do. Clang emits declared variables as alloca, and
```

**Review by jmorse — COMMENTED** [review]
Looks good in principle, with a couple of nits.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (SLTozer) [reply clang/lib/CodeGen/CGDebugInfo.cpp:5787]: To add a little extra to this, it looks like what this code is trying to do is the same as `Instruction::getInsertionPointAfterDef` - would that be a suitable substitution?

Comment (huangjd) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Done.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (huangjd) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Done.

**Review by jmorse — APPROVED** [review]
LGTM, thanks for rapidly rewriting this. I think dereferencing the insertion iterator with `&**` will eventually need addressing with the move away from intrinsics and insertion, but that can be a matter for when the DIBuilder APIs get deprecated / retired.

**Comment by SLTozer** [comment]
Small recommendation that the [patch](https://github.com/llvm/llvm-project/pull/91724) @pogo59 mentioned above has now landed, so before merging this you should rebase and update the test accordingly.

**Comment by dyung** [comment]
@huangjd, I'm not sure the buildbot emails on failure are working, but the test you added seems to be failing on many bots. Can you take a look and revert if you need time to investigate?

- https://lab.llvm.org/buildbot/#/builders/144/builds/163
- https://lab.llvm.org/buildbot/#/builders/46/builds/93
- https://lab.llvm.org/buildbot/#/builders/174/builds/92
- https://lab.llvm.org/buildbot/#/builders/190/builds/108
- https://lab.llvm.org/buildbot/#/builders/23/builds/56

**Comment by huangjd** [comment]
> @huangjd, I'm not sure the buildbot emails on failure are working, but the test you added seems to be failing on many bots. Can you take a look and revert if you need time to investigate?
> 
> * https://lab.llvm.org/buildbot/#/builders/144/builds/163
> * https://lab.llvm.org/buildbot/#/builders/46/builds/93
> * https://lab.llvm.org/buildbot/#/builders/174/builds/92
> * https://lab.llvm.org/buildbot/#/builders/190/builds/108
> * https://lab.llvm.org/buildbot/#/builders/23/builds/56

Test updated https://github.com/llvm/llvm-project/pull/95637

**Pull request merged**
